Tools/Scripts/pgo-profile

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python3
+
+import argparse
+import math
+import subprocess
+import os
+
+def pad(string, max_length):
+    if len(string) > max_length:
+        return string[:max_length - 1] + u"…"
+    return string.ljust(max_length)
+
+
+def shell(command):
+    return subprocess.check_output(command, shell=True, stderr=subprocess.STDOUT).strip()
+
+
+def shortened(name):
+    return name if len(name) < 200 else name[0:99] + u"…" + name[-98:]
+
+
+def assert_directory(d):
+    if os.path.isdir(d):
+        return d
+    else:
+        print("No directory at ", d)
+        exit()
+
+
+def assert_file(f):
+    if os.path.isfile(f):
+        return f
+    else:
+        print("No file at ", f)
+        exit()
+
+
+def summarize_parser(subparsers):
+    parser = subparsers.add_parser("summarize", help="Dumps function names in a given .profraw file, sorted in descending order by function count")
+    parser.add_argument("file", help="Path to the .profraw file")
+    parser.set_defaults(func=summarize)
+    return parser
+
+
+def summarize(args):
+    file = assert_file(args.file)
+
+    lines = shell("xcrun -sdk macosx llvm-profdata show --all-functions --value-cutoff=10 \"{}\" | c++filt -n".format(file)).split(b"\n")
+
+    counts_and_functions = []
+
+    for line_number in range(len(lines)):
+        line = lines[line_number].strip()
+        if line.startswith(b"Function count: "):
+            count = int(line.split()[-1])
+            symbol = lines[line_number - 3].strip()[:-1].decode("utf-8")
+            counts_and_functions.append((count, symbol))
+
+    counts_and_functions.sort(reverse=True)
+    for count, name in counts_and_functions:
+        print(pad(str(count), 15), shortened(name))
+
+
+def merge_parser(subparsers):
+    parser = subparsers.add_parser("merge", help="Merge a pile of *.profraw files into the *.profdata files we can build with.")
+    parser.add_argument("directory", help="Path to the directory containing the *.profraw files")
+    parser.set_defaults(func=merge)
```

Comment:
it's weird to have all these functions act on the argument parser. It's much easier to read what this script does if all argument parser commands are in one place IME. What's the motivation for doing it this way?

Reply:
I figured it was easier to have the definition and use of each of the arguments close together, so you don't have to scroll around to see where the keys are defined

Reply:
My suggestion was to put all of this inside main so it's literally all next to each other.

Reply:
I mean, the definition of an argument like directory, and the use of args.directory. We have to pass in the args dictionary directly, because each function takes a different number of arguments.